### PR TITLE
Answer to Issue #533 + fixes of some bug

### PR DESCRIPTION
**Pull request by StevellM**
I suggest an answer to #533 and also some fixes. I wanted to put into 2 different pull requests but I did a mess with my branches.

### CODE REVIEW

File: src/QuadForm/Quad/ZLattices.jl [hunk 1]
```diff
@@ -491,16 +493,52 @@ function _root_lattice_A(n::Int)
   return z
 end
 
+function _root_lattice_D(n::Int)
+  n < 2 && error("Parameter ($n) for root lattices of type :D must be greater of equal to 2")
```
Comment (simonbrandhorst) [inline src/QuadForm/Quad/ZLattices.jl:497]:
```suggestion
  n < 2 && error("Parameter ($n) for root lattices of type :D must be greater or equal to 2")
```

File: src/QuadForm/Quad/ZLattices.jl [hunk 2]
```diff
@@ -491,16 +493,52 @@ function _root_lattice_A(n::Int)
...
+  if n == 2
+    G = matrix(ZZ, [2 0 ;0 2])
+  elseif n == 3
+    return _root_lattice_A(n)
+  else
+    G = zero_matrix(ZZ, n, n)
+    G[1,3] = G[3,1] = -1
+    for i in 1:n
+      G[i,i] = 2
+      if i in 2:n-1
```
Comment (simonbrandhorst) [inline src/QuadForm/Quad/ZLattices.jl:507]:
```suggestion
      if 2 <= i <= n-1
```

File: test/QuadForm/Quad/ZLattices.jl [hunk 3]
```diff
@@ -347,6 +341,39 @@ end
   @test ambient_space(N) === ambient_space(L)
   @test rank(N) == 0
   @test basis_matrix(invariant_lattice(L, identity_matrix(QQ, 2))) == basis_matrix(L)
+  
+  randlist = rand(2:70,10)
```
Comment (simonbrandhorst) [inline test/QuadForm/Quad/ZLattices.jl:345]:
```suggestion
  randlist = rand(2:20,10)
```
There is no need to test so many or big lattices. If it works for `2:10` it works.

File: test/QuadForm/Quad/ZLattices.jl [hunk 4]
```diff
@@ -347,6 +341,39 @@ end
...
+  L = [root_lattice(:D,i) for i in randlist]
+  @test any(l -> discriminant(l) == 4, L)
+  @test any(iseven, L)
+  @test any(l -> norm(l) == 2, L)
+
+  L = root_lattice(:D,3)
+  #Problem here: equality of lattices depends on 'strict' equality of their ambient
```
Comment (simonbrandhorst) [inline test/QuadForm/Quad/ZLattices.jl:352]: It is intended that only lattices in the same ambient space compare. 
Otherwise `isisometric` can be used.

**Review by simonbrandhorst — CHANGES_REQUESTED** [review]

**Comment by simonbrandhorst** [comment]
@thofma Please confirm the changes on the associative algebra part.

File: src/AlgAssRelOrd/Ideal.jl [hunk 5]
```diff
@@ -1432,11 +1432,13 @@ end
 # factor.
 # Always considers I as an ideal of its left order.
 function factor(I::AlgAssRelOrdIdl)
+  @assert !iszero(I)
   O = left_order(I)
   @assert ismaximal(O)
+  basis_pmatrix(I) != basis_pmatrix(1*O) || error("I must be proper")
```
Comment (thofma) [inline src/AlgAssRelOrd/Ideal.jl:1438]:
```suggestion
  I != 1*O || error("I must be proper")
```

File: src/QuadForm/Quad/ZLattices.jl (hunk 2 above)
Comment (StevellM) [reply src/QuadForm/Quad/ZLattices.jl:507]: Now that we are moving equality of lattices, should I keep the definition of `D3` like this, being equal to `A3` or should I make standard like `D4`,`D5`, and so on ? In this case `D3` and `A3` will be isometric but not equal.

Comment (simonbrandhorst) [reply src/QuadForm/Quad/ZLattices.jl]: It doesn't really matter I think. So you can keep it. There is also an elegant description of the `A,D,E` root lattices as a sublattice of the standard euclidean space of dimension `n+1`. But lets keep the ambient keyword for later :-).

**Pull request merged**